### ISSUE TITLE
builtins.fetchGit requires git

### ISSUE THREAD

**Issue by zoranbosnjak** [NixOS/nix#3533]
When using `builtins.fetchGit`, the `nix-build` command won't work if:
* `git` program is not installed on the system
* or is installed but is too old (flag `-C <path>` is not supported)

I am testing on non-nix linux distributions.

This does not look right. The `builtins.fetchGit` obviously depends on the correct version of the `git` program. I would expect that `nix` would build/download `git` automatically when required and the correct version is used, even if some other version is installed on the system.

**Comment by worldofpeace** [comment]
What really should be done, if git isn't found, it should bail out with a helpful error.
Same with fetchTarball.

**Comment by LnL7** [comment]
Pretty outdated now, but I tried to improve the errors in https://github.com/NixOS/nix/pull/2498

**Comment by mkenigs** [comment]
Fixing could be done with something similar to https://github.com/NixOS/nix/pull/3565

**Comment by mkenigs** [comment]
@edolstra was taking a look at this, any suggestion on how to make InstallableFlake accessible in libfetcher like you mentioned in https://github.com/NixOS/nix/pull/3565#discussion_r420658057?

**Comment by balsoft** [comment]
Shouldn't this be fixed by just wrapping Nix itself with git, tar and other tools needed at runtime?

**Comment by rnhmjoj** [comment]
I agree. Is Nix a special case? Generally if something needs to call a binary, it should be wrapped or patched to call the binary from a nix store path.

**Comment by hectorj** [comment]
Hi!

I have the same problem today: I can't use `builtins.fetchGit` on my NixOS VM because it does not find git.

My workaround is running ` nix shell nixpkgs#git -c nixos-rebuild build` instead of just `nixos-rebuild build`.

It works but it is does not seem ideal.

**Comment by xokdvium** [comment]
There's the option of falling back to libgit2 with a warning. Nix still relies on the git binary for credential helpers and fetch performance, but at least trying to do something sounds better than not being able to fetch anything at all.
